### ISSUE TITLE
Unserializable requests slow down the engine?

### ISSUE THREAD

**Issue by Prometheus3375** [scrapy/scrapy#6119]
# Description

We are using a class to generate company registration numbers (CRNs) and then yield request with them. If 1000 consecutive requests fail, then class instances exit prematurely. To correctly satisfy this condition, we changed class, so from iterable it became iterator. This rendered its instances to be unpickleable, so all requests stared to go into RAM queue instead of disk queue.

For some reason this caused a massive slow down over time. From initial 8k requests per minute it was reducing to 2k. Here is the graph.

![image](https://github.com/scrapy/scrapy/assets/35541026/4c848e46-12ad-4ff6-87f0-e8fe2f5df8ad)

10/10 - 10/12 is before the change iterable -> iterator was introduced. The sum of 404 and 200 responses is always near 8k.
10/12 - 10/14 is incomprehensible, so I am skipping it.
10/14 - 10/20 -- here you can see spikes of 404 and then gradual slow down without increase of 202. Every spike is restart of the spider.
10/20 - present -- I applied a fix which I will discuss later. Here the situation is starting to look similar to 10/10 - 10/12 as it should be.

Here is also RAM usage for that time.
![image](https://github.com/scrapy/scrapy/assets/35541026/ab526c64-b050-4915-94f8-2672e616c484)
As you can observe, 10/14 - 10/20 have the same RAM usage as 10/10 - 10/12 and 10/20 - present.

# Steps to Reproduce

I designed a sample spider which reproduces the issue.
[spider.txt](https://github.com/scrapy/scrapy/files/13051998/spider.txt)
Start it with next command.
```
crawl _test -s JOBDIR="_state"
```
There is no graceful shutdown in the example, so kill it after 2 hours.

I ran it for 2 hours and got next logs. Some notes about them:
1. There was an error with item/minute calculation, it is fixed in attached spider code.
2. About used 3rd party stuff:
   - WunderStats is the same as CoreStats.
   - WunderDupeFilter just adds buffering to `requests.seen` and is unused in this test because all requests have `dont_filter` flag.
   - Spider loader just loads spider, should not affect the test.
   - RandomUserAgentMiddleware and RetryUserAgentMiddleware just substitute user-agent with a random one, everything else is the same as in the default Scrapy middlewares.
 3. Any 3rd party stuff seem not to cause slow down, because they all are still in use, but the issue is fixed now.

[log.txt](https://github.com/scrapy/scrapy/files/13051813/log.txt)
As the log provides, the speed went from 10k on average to 7k. Once I restarted the spider, I again had 10k speed.

# Expected behavior

Constant average response/minute speed.

# Actual behavior

Response/minute is going down after an hour and continues to decrement later. Always reproduces.

# Versions

```
Scrapy       : 2.11.0
lxml         : 4.9.3.0
libxml2      : 2.10.3 
cssselect    : 1.2.0  
parsel       : 1.8.1  
w3lib        : 2.1.2
Twisted      : 22.10.0
Python       : 3.11.5 (tags/v3.11.5:cce6ba9, Aug 24 2023, 14:38:34) [MSC v.1936 64 bit (AMD64)]
pyOpenSSL    : 23.2.0 (OpenSSL 3.1.3 19 Sep 2023)
cryptography : 41.0.4
Platform     : Windows-10-10.0.19044-SP0
```

# The fix

Make all requests serializable. In the provided spider pass index of iterator instead of the iterator itself. The same approach we are using in the production now.

Changed methods of the spider:

```python
    def start(self, /):
        it = ((n, gen, crn) for n, gen in enumerate(self.generators) for crn in gen)
        for n, generator, crn in it:
            if crn is None:
                yield Request('data:,', self.__start, dont_filter=True, errback=self.__start)
                self.logger.info(
                    f'Generator {generator} returned None; '
                    f'generation will proceed in 60 seconds'
                    )
                return

            self.latest_crn = crn
            yield Request(
                'https://example.com',
                self.check_crn_status,
                dont_filter=True,
                errback=self.check_crn_status_failed,
                cb_kwargs=dict(crn=crn, generator_n=n),
                )

    def check_crn_status(self, _: Response, /, *, crn: str, generator_n: int):
        exists = bools[0]
        self.generators[generator_n].process_crn(crn, exists)
        if exists:
            yield {
                '_id': crn,
                'ts':  datetime.now(timezone.utc),
                }

        bools.rotate(-1)

    def check_crn_status_failed(self, failure: Failure, /):
        # noinspection PyUnresolvedReferences
        request: Request = failure.request
        kwargs = request.cb_kwargs
        crn: str = kwargs['crn']
        generator_n: int = kwargs['generator_n']
        self.logger.error(f'Existence check for {crn!r} has failed -- {failure.getErrorMessage()}')
        self.generators[generator_n].process_crn(crn, False)
```

I applied these changes and started the spider for 2 hours.
[log_fix.txt](https://github.com/scrapy/scrapy/files/13053235/log_fix.txt)
The result: no slow down, the average speed was 9.5k-8.5k all the time. There is only one record with 7k speed compared to the previous log where 5k records are present.

**Comment by Prometheus3375** [comment]
I did some additional testing. I designed a more simpler spider to check the statement "Unserializable requests slow down the engine".

[spider.txt](https://github.com/scrapy/scrapy/files/13063993/spider.txt)
[settings.txt](https://github.com/scrapy/scrapy/files/13063990/settings.txt)

I started this spider in parallel with a serializable version (where `it = [i for i in range(10)]`) on cluster and here are the results.

Serializable vs. unserializable.
![image](https://github.com/scrapy/scrapy/assets/35541026/2b1ea18d-ea6a-4476-8175-e7fd04b85b02)
![image](https://github.com/scrapy/scrapy/assets/35541026/cdb2fc6a-4a4c-480a-8973-fa87fb87ff1d)

Unserializable was slower, but there was no slowdown at all, the speed was constant. Thus, the combination of generator design and inability to serialize generators produces the issue.

I returned back to the spider from the original comment, but changed request URLs from `https://example.com` to `data:,` to eliminate network issues from the contribution. I also disabled retry, since it is unneeded, disabled 3rd party user agent and enabled provided by the Scrapy one. I again started both versions in parallel.

Serializable vs. unserializable.
![image](https://github.com/scrapy/scrapy/assets/35541026/d9dd8228-df3c-45be-8f83-9aed284439fc)
![image](https://github.com/scrapy/scrapy/assets/35541026/88d81e6e-cd01-4014-85a7-328f9351f014)

Serializable version was also slowing down, but the slow down was not so drastic as in the case with unserializable. The effect is very noticeable: serializable version scrapped twice more than unserializable one within the same period of time. At the end, serializable version consumed ~100 MB of RAM, while unserializable ~85 MB; this is expected, because serializable just stored more indexes inside set `MyIterator.all_existing`.

I am planning to make another test with a simpler generator structure to check whether passing generator in-use causes the issue.

**Comment by Prometheus3375** [comment]
I did two more tests with modified simpler spider and got the same results, so I again decided to use the original spider, but w/o set `MyIterator.all_existing`. And once I started it, I immediately realized that it causes the issue.

`MyIterator` does not have methods that tells `pickle` how to make serialization, so it uses a default one. I assume, it uses `dir` to get the list of attributes. `dir` returns attribute names in alphabetical order; thus, `all_existing` comes before any other attribute of `MyIterator` objects. This causes `pickle` to serialize this set at first; eventually, `pickle` encounters attribute `iterator` and fails to serialize.

The more time passes, the bigger `all_existing` becomes, the longer it takes for `pickle` to serialize the set before inevitable fail, the slower crawling speed becomes.

In addition, I also realized that passing mutable objects to `cb_kwargs` when `JOBDIR` is set is a DUMB idea.

```python
class TestSpider(Spider):
    name = 'test'

    custom_settings = {
        'JOBDIR': '.state',
        }

    def start_requests(self, /):
        my_set = set()
        self.logger.info(f'Set ID: {id(my_set)}')
        return Request('data:,', self.check, cb_kwargs={'my_set': my_set})

    def check(self, _, /, *, my_set: set):
        self.logger.info(f'Set ID in callback: {id(my_set)}')
```

In the example above, the sets are different. Any changes made to `my_set` in `check` make no effect to `my_set` defined in `start_requests`. Guess why my code was not working as intended initially. 😅

---

All in all, the issue is not the Scrapy fault. However, some improvements can be made:

- [The documentation about jobs](https://docs.scrapy.org/en/latest/topics/jobs.html) does not include the fact that the objects passed to the callback are different from the object passed to dictionary `cb_kwargs`. One of they way describing it:
  (after the paragraph about debug setting)

  > Keep in mind, that deserialization of requests creates deep copies of the objects passed to 
  dictionary `cb_kwargs`. Meaning, that any mutable object inside `cb_kwargs` (directly or as a part 
  of other object) does not make its way to the callback; the callback receives its exact copy. Any 
  changes made to the copy does not affect the original object.

- If `JOBDIR` is set, emit a warning, if `cb_kwargs` holds common mutable objects as values. Common mutable objects are lists, dicts, sets, bytearrays, arrays from `array` module (this can be not a complete list).
